pkg/controllers/nodepool/counter/provisioner_test.go

```diff
@@ -0,0 +1,107 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package counter_test
+
+import (
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider/fake"
+	"github.com/aws/karpenter-core/pkg/test"
+	. "github.com/aws/karpenter-core/pkg/test/expectations"
+)
+
+var _ = Describe("Provisioner Counter", func() {
+	BeforeEach(func() {
+		cloudProvider.InstanceTypes = fake.InstanceTypesAssorted()
+		provisioner = test.Provisioner()
+		instanceType := cloudProvider.InstanceTypes[0]
+		machine, node = test.MachineAndNode(v1alpha5.Machine{
+			ObjectMeta: metav1.ObjectMeta{Labels: map[string]string{
+				v1alpha5.ProvisionerNameLabelKey: provisioner.Name,
+				v1.LabelInstanceTypeStable:       instanceType.Name,
+			}},
+			Status: v1alpha5.MachineStatus{
+				ProviderID: test.RandomProviderID(),
+				Capacity: v1.ResourceList{
+					v1.ResourceCPU:    resource.MustParse("100m"),
+					v1.ResourcePods:   resource.MustParse("256"),
+					v1.ResourceMemory: resource.MustParse("1Gi"),
+				},
+			},
+		})
+		machine2, node2 = test.MachineAndNode(v1alpha5.Machine{
+			ObjectMeta: metav1.ObjectMeta{Labels: map[string]string{
+				v1alpha5.ProvisionerNameLabelKey: provisioner.Name,
+				v1.LabelInstanceTypeStable:       instanceType.Name,
+			}},
+			Status: v1alpha5.MachineStatus{
+				ProviderID: test.RandomProviderID(),
+				Capacity: v1.ResourceList{
+					v1.ResourceCPU:    resource.MustParse("500m"),
+					v1.ResourcePods:   resource.MustParse("1000"),
+					v1.ResourceMemory: resource.MustParse("5Gi"),
+				},
+			},
+		})
+		ExpectApplied(ctx, env.Client, provisioner)
+		ExpectReconcileSucceeded(ctx, provisionerInformerController, client.ObjectKeyFromObject(provisioner))
+		ExpectReconcileSucceeded(ctx, provisionerController, client.ObjectKeyFromObject(provisioner))
+		provisioner = ExpectExists(ctx, env.Client, provisioner)
+	})
+
+	It("should increase the counter when a new nodes are created", func() {
+		ExpectApplied(ctx, env.Client, node, machine)
+		ExpectMakeNodesMachinesInitializedAndStateUpdated(ctx, env.Client, nodeController, machineController, []*v1.Node{node}, []*v1alpha5.Machine{machine})
+
+		ExpectReconcileSucceeded(ctx, provisionerController, client.ObjectKeyFromObject(provisioner))
+		provisioner = ExpectExists(ctx, env.Client, provisioner)
+
+		// Should equal both the machine and node capacity
+		Expect(provisioner.Status.Resources).To(BeEquivalentTo(machine.Status.Capacity))
```

Comment:
I think generally speaking, the Machine and Node capacity won't match. Maybe to make this more realistic, we should:
1. Add a machine
2. Check the provisioner status
3. Register a node and initialize it
4. Check the provisioner status to ensure it now matches the node

Reply:
Good callout! Added another test here.